design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,438 @@
+
+A second CRD `FirmwareSchema` will added to store schema data (aka BIOS
+Registry) received from Ironic via the BIOS API. This schema provides
+additional info about each BIOS setting along with limits that can be used
+both by a user when changing the settings or as a part of input validation.
+The BIOS Registry support was recently added to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681).
+This schema is the same for each vendor's model so it will not need to be
+updated when stored. More info for the BIOS registry data can be found in the
+[Redfish Registry](https://redfish.dmtf.org/schemas/v1/AttributeRegistry.v1_3_5.json)
+
+The schema fields returned via the BIOS API include:
+
+- Type (e.g. enum, string, integer, boolean, or password)
+- ReadOnly (a boolean indicating whether the attribute is read only)
+- Unique (indicating the value is specific to the host where it was
+  retrieved)
+- Allowable values for enum type attributes
+- Minimum and maximum values (for numeric attributes)
+- Minimum and maximum character lengths (for string attributes)
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. The baremetal-operator will manage the data as follows:
+
+- The node first transitions to manageable during the `Registration` state, so
+  at the end of that state `settingsData` will be populated.
+- Settings can be updated during the `Preparing` state, so at the end of that
+  state the settings will also be retrieved and used to update
+  `settingsData`.
+
+Upon retrieving the BIOS Settings from Ironic, the baremetal-operator will
+copy the name/value pairs to the `vendorSettings` field in
+`HostFirwareSettingsSpec` if that field is empty. Settings that have the
+`read-only` flag set will not be copied as these settings cannot be written
+to the BMC. Likewise, settings that have the `unique` flag set will also not
+be copied, these are settings like serial numbers or product ids for example,
+that should not be written to other servers.
+
+A user can update `vendorSettings` to the desired values for the
+settings using the limits from the `FirmwareSchema` section as a guide.
+
+The baremetal-operator will detect when `vendorSettings` changes by comparing
+the name/value pairs to `settingsData`. When a change is detected, the BMO
+will add the new values to the Ironic clean-steps API in the Preparing state,
+when building the manual clean steps. Before adding it though,
+the BMO will do validation checks on the new values using the schema. For
+example, integer types will checked against minimum and maximum parameters and
+enumeration types will be compared against allowable values. The validation
+webhook can be leveraged here to log a message regarding the failure; the
+invalid BIOS Setting will not be added to clean-steps.
```

Comment:
I think this paragraph might be based on a misunderstanding of what a validating webhook does.

The webhook sits in the API pipeline between the input and the data store, so it allow us to reject an invalid update before it is stored in the resource. Ideally we should do this so that the user knows immediately if they submit an invalid change.

Without a webhook we end up with invalid config stored in the object, and then the controller has to decide what to do with it given that it's too late to reject it. Since we'd have to put the Host in an error state anyway, and retry after the user has fixed it, I agree there'd probably not be much point in doing a partial application.

Reply:
Updated to state that validity checks will be done when building the manual clean steps and will return a failure for any failures, preventing partial configurations.  

Also indicated that a validation webhook will be done as a follow-on, to catch validation errors before storing the data in a resource.